### PR TITLE
Build libbcc and system-probe is a dedicated build environment

### PR DESCRIPTION
**Pull request by L3n41c**
### What does this PR do?

This PR adds [iovisor/bcc](https://github.com/iovisor/bcc) as a new runtime dependency of `system-probe`.
This is a dependency that comes from [iovisor/gobpf](https://github.com/iovisor/gobpf) which is already vendored in `system-probe`, but which is updated in this PR. (Strictly speeking, we are using the [datadog fork of it](https://github.com/DataDog/gobpf)).
[iovisor/bcc](https://github.com/iovisor/bcc) requires a build environment which is more up-to-date than the `centos:6` and `debian:wheezy` environment that are currently used to build the agent.
Previous attempts to build [iovisor/bcc](https://github.com/iovisor/bcc) on the existing build environments failed because of the compiler being too old (lack of C++14 support), the Linux kernel header being too old and the glic headers being too old (`netlink.h` missing some constants)

So, the native library [iovisor/bcc](https://github.com/iovisor/bcc) is built in a dedicated more up-to-date build environment created by DataDog/datadog-agent-buildimages#25.
[iovisor/gobpf](https://github.com/iovisor/gobpf) which is vendored in `system-probe` is also requiring up-to-date glibc headers. That’s why it is also built in dedicated build environment.

Both `libbcc` and `system-probe` are then copied to the final image, in the omnibus tree (`/opt/datadog-agent/embedded`) where omnibus validates that all symbols found in this library can be resolved in libraries that are also found in the omnibus tree.

### Motivation

This the first part of #4619 which aims at introducing an eBPF-based check that would perform runtime eBPF compilation.
Current implementation of `system-probe` does a one-time compile-time eBPF code compilation and then do runtime kernel structures offset guessing.
This approach has some limitations that we would like to overcome by doing runtime eBPF code compilation.
[iovisor/bcc](https://github.com/iovisor/bcc) and its GO wrapper [iovisor/gobpf](https://github.com/iovisor/gobpf) are providing a convenient framework to implement runtime eBPF code compilation.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -512,6 +563,38 @@ run_dogstatsd_arm_size_test:
...
+.system-probe_build_common:
+  before_script:
+    # Hack to work around the cloning issue with arm runners
+    - mkdir -p $GOPATH/src/github.com/DataDog
+    - '[[ "$ARCH" == arm64 ]] && cp -R $GOPATH/src/github.com/*/*/DataDog/datadog-agent $GOPATH/src/github.com/DataDog'
+    - cd $SRC_PATH
+    - inv -e deps --verbose --dep-vendor-only
+    # Retrieve libbcc from S3
+    - $S3_CP_CMD $S3_ARTIFACTS_URI/libbcc-$ARCH.tar.xz /tmp/libbcc.tar.xz
+    - tar -xvf /tmp/libbcc.tar.xz -C /usr
```
Comment (albertvaka) [inline .gitlab-ci.yml]: This is a bit dirty :/

File: omnibus/config/software/libbcc.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https:#www.datadoghq.com/).
+# Copyright 2016-2020 Datadog, Inc.
+
+name 'bcc'
+
+dependency 'libelf'
+
+build do
+  command "#{ENV['S3_CP_CMD']} #{ENV['S3_ARTIFACTS_URI']}/libbcc-#{ENV['PACKAGE_ARCH']}.tar.xz /tmp/libbcc.tar.xz"
```
Comment (albertvaka) [inline omnibus/config/software/libbcc.rb]: What would happen if a user wants to build the system-probe (ie: outside the CI)? I think it will fail because `S3_ARTIFACTS_URI` is not defined?

Maybe we can treat this new dependency the same way we treat `jmxfetch`, for which we specify a pre-built version here [1] and download it in omnibus. This way, if a user does a custom build, it will still pick up whatever version is specified. Plus, it will be re-built only when there are changes.

[1] https://github.com/DataDog/datadog-agent/blob/master/release.json#L6

Comment (albertvaka) [reply omnibus/config/software/libbcc.rb]: Actually... is there any good reason not to treat the entire system-probe this way, like we do with jmxfetch? @olivielpeau @sunhay What's the benefit of having it live in the same repo?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -170,6 +173,51 @@ fail_on_non_triggered_tag:
...
+
+# build libbcc
+.build_libbcc_common:
+  stage: deps_build
+  before_script: []
+  script:
+    - git clone -b "$BCC_VERSION" --depth=1 https://github.com/iovisor/bcc.git /tmp/bcc
+    - mkdir /tmp/bcc/build
+    - cd /tmp/bcc/build
+    - cmake .. -DCMAKE_INSTALL_PREFIX=/opt/libbcc -DCMAKE_EXE_LINKER_FLAGS='-Wl,-rpath,/opt/datadog-agent/embedded/lib' -DCMAKE_SHARED_LINKER_FLAGS='-Wl,-rpath,/opt/datadog-agent/embedded/lib'
```
Comment (leeavital) [inline .gitlab-ci.yml]: Could this be built into a new build image? Doing this on every build introduces a build time dependency on github and is slow (compiling bcc anecdotally takes a minute on my laptop)

File: omnibus/config/projects/agent.rb [hunk 4]
```diff
@@ -157,6 +157,11 @@
 # Datadog agent
 dependency 'datadog-agent'
 
+# System-probe
+if linux?
+  dependency 'system-probe'
+end
+
```
Comment (leeavital) [inline omnibus/config/projects/agent.rb]: Why is this required now ?

File: omnibus/config/projects/puppy.rb [hunk 5]
```diff
@@ -124,6 +124,12 @@
 if windows?
   dependency 'datadog-agent-finalize'
 end
+
+# System probe
+if linux?
+  dependency 'system-probe'
+end
+
```
Comment (leeavital) [inline omnibus/config/projects/puppy.rb]: I'm not too familiar with the puppy build, but is it an issue for the process-agent to be missing ?

File: omnibus/config/software/libbcc.rb (hunk 2 above)
Comment (leeavital) [reply omnibus/config/software/libbcc.rb]: We used to have it this way, and merged it when the process-agent and trace-agents were being merged into the single repo.

Having them in the same repo has some advantages:
- releasing is easier as the process for pulling a new system-probe into the agent was (tag system-probe >   agent PR to pull new version in > tag agent) 
-  dependency management is easier since the system-probe pulls in some agent packages (mostly utilities and stuff related to config)

File: tasks/system_probe.py [hunk 6]
```diff
@@ -256,7 +256,7 @@ def build_object_files(ctx, install=True):
 
     for d in linux_headers:
         for s in subdirs:
-            flags.extend(["-I", os.path.join(d, s)])
+            flags.extend(["-isystem", os.path.join(d, s)])
```
Comment (leeavital) [inline tasks/system_probe.py]: why does this need to change?

File: omnibus/config/software/libbcc.rb (hunk 2 above)
Comment (albertvaka) [reply omnibus/config/software/libbcc.rb]: It makes sense if it depends on shared files.

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -170,6 +173,51 @@ fail_on_non_triggered_tag:
...
+# deps_build
+#
+#
+
+empty:
+  stage: deps_build
+  tags: [ "runner:main", "size:large" ]
+  before_script: []
+  script:
+    - echo "This is an empty step that can be used as a dependency for jobs that depend on nothing"
```
[the remark this answers is not in the document]
Comment (L3n41c) [reply .gitlab-ci.yml]: Indeed !
I had to introduce this `empty` step before `fail_on_non_triggered_tag` was introduced.
Let me fix that !

File: omnibus/config/software/libbcc.rb (hunk 2 above)
Comment (L3n41c) [reply omnibus/config/software/libbcc.rb]: The core agent and system-probe are, in the end, packaged together in a single package/docker image. So, they share the same lifecycle.
In a followup PR (#4619), I introduce communication between the two.
The fact that they are both hosted in the same repository ease that work since the data structures that are exchanged between the two can be defined at only one place.

File: tasks/system_probe.py (hunk 6 above)
Comment (L3n41c) [reply tasks/system_probe.py]: This is indeed somewhat unrelated.
This allows to remove all the `#pragma clang diagnostic ignored …` that there are around `#include <linux/ptrace.h>` and `#include <net/sock.h>` in `pkg/ebpf/c/tracer-ebpf.c`.
Indeed, those two included files are system headers and they may generate compiler warnings, depending on the warnings that we have enabled few lines above in the same file.
I understand that we want aggressive warning checks for our own code (`tracer-ebpf.c`) but we won’t fix the system headers to make them compliant with our own warning options.
That’s why there used to be some `#pragma clang diagnostic ignored …` lines.
But when I moved to a more up-to-date build environment, the system headers were different and were raising different warnings.
Rather than adding some new `#pragma`, it is cleaner to just tell the compiler that those paths are containing system headers and not user headers.

File: omnibus/config/projects/agent.rb (hunk 4 above)
Comment (L3n41c) [reply omnibus/config/projects/agent.rb]: That’s because I removed the `invoke system-probe.build` invocation that there was in `omnibus/config/software/datadog-agent.rb` line 124.

File: .gitlab-ci.yml (hunk 3 above)
Comment (L3n41c) [reply .gitlab-ci.yml]: All the dependencies of the agent are currently built by omnibus in the agent pipeline.
Here is an example of an agent pipeline logs: https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/25136857
We can see behind the:
```
[Builder: procps-ng] …
[Builder: zlib] …
[Builder: cacerts] …
[Builder: xproto] …
[Builder: util-macros] …
[Builder: libiconv] …
… many more …
```
the logs of the build of many 3rd party dependencies.

The specificity of `libbcc` is that it requires a different build environment. So, it is built in its own job.
But if we want to build dependencies in a distinct pipeline in order to speed up the agent one, shouldn’t we do that for all the 3rd party dependencies and not only for `libbcc` ?

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -268,16 +318,15 @@ run_tests_rpm-x64-py3:
 # run tests for eBPF code
 run_tests_ebpf:
   stage: source_test
-  # TODO(processes): change this to use the generic builder image when we use the same version of Go in the
-  # system-probe and the agent, or when `system-probe.test` can receive which Go version to use as an argument.
-  # Currently this is using the ebpf image so system-probe tests run in the same  Go version used to build it (1.10).
-  # Note that this image isn't used to build the system-probe: when building it, we specify Go version to use
-  # through an argument to the system-probe.build invoke task.
-  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/ebpf:go1.10.1
+  needs: ["build_libbcc_x64"]
```
[the remark this answers is not in the document]
Comment (L3n41c) [reply .gitlab-ci.yml]: Done !
I’ve put `"fail_on_non_triggered_tag"` on every single `needs:` directive.

File: .gitlab-ci.yml (hunk 1 above)
Comment (L3n41c) [reply .gitlab-ci.yml]: This was to have `inv system-probe.test` able to find `libbcc` like any other system library it is depending on.
Note that this step is not building any delivery. So that has no implication.
Would you prefer to have `libbcc` unarchived in a dedicated directory (`/opt/libbcc` or `/opt/datadog-agent/embedded`) and to pass appropriate `CGO_CFLAGS` `GCO_LDFLAGS` to `inv system-probe.test` ?

File: omnibus/config/projects/puppy.rb (hunk 5 above)
Comment (albertvaka) [reply omnibus/config/projects/puppy.rb]: We don't include trace nor processes in puppy, so following that we shouldn't include system-probe either?

File: .gitlab-ci.yml (hunk 1 above)
Comment (albertvaka) [reply .gitlab-ci.yml]: The cleanest solution I can think of is to also use omnibus to build the system-probe, so omnibus downloads this and puts it into the omnibus build prefix... but it's probably a pain in the ass to implement :P

File: .gitlab-ci.yml (hunk 3 above)
Comment (L3n41c) [reply .gitlab-ci.yml]: Because of the way the dependencies are described, the new `build_libbcc_{arm64,x64}` jobs are run in parallel to the `run_tests_*` ones.
Both sets of jobs are required to move to the `Binary_build` stage.
And `build_libbcc_{arm64,x64}` jobs are faster than the `run_tests_*` ones.
So, in the end, adding those extra jobs doesn’t change the total time it takes for the whole pipeline to complete because it runs in parallel to existing jobs and they are not on the critical path.

File: omnibus/config/projects/puppy.rb (hunk 5 above)
Comment (L3n41c) [reply omnibus/config/projects/puppy.rb]: Good catch. I’ve removed `system-probe` from the puppy build.

File: omnibus/config/software/system-probe.rb [hunk 9]
```diff
@@ -8,6 +8,7 @@
 dependency 'libbcc'
 
 build do
-  command "#{ENV['S3_CP_CMD']} #{ENV['S3_ARTIFACTS_URI']}/system-probe.#{ENV['PACKAGE_ARCH']} #{install_dir}/embedded/bin/system-probe"
-  command "chmod 755 #{install_dir}/embedded/bin/system-probe"
```
Comment (albertvaka) [inline omnibus/config/software/system-probe.rb]: chmod no longer needed?

File: .gitlab-ci.yml [hunk 10]
```diff
@@ -601,7 +601,10 @@ build_system-probe-arm64:
     # Artifacts and cache must live within project directory but we run omnibus in a neutral directory.
     # Thus, we move the artifacts at the end in a gitlab-friendly dir.
     # Use --skip-deps since the deps are installed by `before_script`.
-    - inv -e agent.omnibus-build --release-version "$RELEASE_VERSION" --major-version "$AGENT_MAJOR_VERSION" --python-runtimes "$PYTHON_RUNTIMES" --base-dir $OMNIBUS_BASE_DIR ${USE_S3_CACHING} --skip-deps
+    - $S3_CP_CMD $S3_ARTIFACTS_URI/system-probe.${PACKAGE_ARCH} /tmp/system-probe
+    - chmod 755 /tmp/system-probe
+    - $S3_CP_CMD $S3_ARTIFACTS_URI/libbcc-${PACKAGE_ARCH}.tar.xz /tmp/libbcc.tar.xz
+    - SYSTEM_PROBE_BIN=/tmp/system-probe LIBBCC_TARBALL=/tmp/libbcc.tar.xz inv -e agent.omnibus-build --release-version "$RELEASE_VERSION" --major-version "$AGENT_MAJOR_VERSION" --python-runtimes "$PYTHON_RUNTIMES" --base-dir $OMNIBUS_BASE_DIR ${USE_S3_CACHING} --skip-deps
```
Comment (albertvaka) [inline .gitlab-ci.yml]: Small nit: `SYSTEM_PROBE_BIN` and `LIBBCC_TARBALL` are passed as env vars, while everything else is passed as arguments to invoke (which in turn converts to env vars [1]). To be consistent (and also making these parameters more easily discoverable), can you change the env vars to arguments?

[1] https://github.com/DataDog/datadog-agent/blob/master/tasks/agent.py#L404

File: omnibus/config/software/system-probe.rb (hunk 9 above)
Comment (L3n41c) [reply omnibus/config/software/system-probe.rb]: That’s because I moved the `chmod` to the `.gitlab-ci.yml` job just after the `s3 cp` command.
The thing is that it’s S3 which is loosing the execution bit.
If the user generates `system-probe` in one container and uses it in another one to build the agent and uses a docker volume to share `system-probe` between the two, no `chmod` is needed anywhere.
It’s really the S3 copy which is loosing it. That’s why I kept the `chmod` close to the `s3 cp` command and moved the two out of omnibus and put both in `.gitlab-ci.yml`.

Comment (albertvaka) [reply omnibus/config/software/system-probe.rb]: Make sense now 👍

File: .gitlab-ci.yml (hunk 10 above)
Comment (L3n41c) [reply .gitlab-ci.yml]: Indeed !
Change implemented.
Thanks for the advice.

File: omnibus/config/software/system-probe.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# Unless explicitly stated otherwise all files in this repository are licensed
+# under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https:#www.datadoghq.com/).
+# Copyright 2016-2020 Datadog, Inc.
+
+name 'system-probe'
+
+dependency 'libbcc'
```
Comment (albertvaka) [inline omnibus/config/software/system-probe.rb]: Should we comment this out for now until it's used? So we don't ship libbcc in 7.18 since it's not used.

Comment (L3n41c) [reply omnibus/config/software/system-probe.rb]: Good idea !
I’ve just commented the `libbcc` dependency out.

**Review by albertvaka — APPROVED** [review]
LGTM.

File: .gitlab-ci.yml [hunk 12]
```diff
@@ -170,6 +173,44 @@ fail_on_non_triggered_tag:
...
+  script:
+    - git clone -b "$BCC_VERSION" --depth=1 https://github.com/iovisor/bcc.git /tmp/bcc
+    - mkdir /tmp/bcc/build
+    - cd /tmp/bcc/build
+    - cmake .. -DCMAKE_INSTALL_PREFIX=/opt/libbcc -DCMAKE_EXE_LINKER_FLAGS='-Wl,-rpath,/opt/datadog-agent/embedded/lib' -DCMAKE_SHARED_LINKER_FLAGS='-Wl,-rpath,/opt/datadog-agent/embedded/lib'
+    - make -j 4 #"$(nproc)"
+    - make install
+    - cd /opt/libbcc
+    - rm share/bcc/introspection/bps
+    - cp $(ldd lib/libbcc.so | awk '$1 ~ /^libtinfo/ {system("dirname " $3)}')/libtinfo* lib
```
Comment (leeavital) [inline .gitlab-ci.yml]: Can you add a comment explaining this part ?

**Review by leeavital — APPROVED** [review]
This approach looks good to me, though we should discuss embedding libbcc with security. I'm not terribly familiar with the packaging bits, I'll defer to @albertvaka for those

File: omnibus/config/software/libelf.rb [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+build do
+  command %q(patch -p 1 <<"EOF"
+--- elfutils-0.178/src/elfclassify.c	2019-11-26 22:48:42.000000000 +0000
++++ elfutils-0.178.patched/src/elfclassify.c	2020-01-28 09:22:28.066520000 +0000
+@@ -827,7 +827,10 @@
+       break;
+     case do_print0:
+       if (checks_passed == flag_print_matching)
++#pragma GCC diagnostic push
```
[the remark this answers is not in the document]
Comment (L3n41c) [reply omnibus/config/software/libelf.rb]: This is most probably due a combination of the version of GCC that we use, the version of the glibc headers and the compilation options used by `libelf`.
Without that `#pragma`, `libelf` compilation fails with with an error about the fact that the return value of `fwrite` is ignored.
